### ISSUE TITLE
Dockerized LaunchedURLClassLoader fails to access package private class

### ISSUE THREAD

**Issue by timtebeek** [spring-projects/spring-boot#17811]
Sample application here: https://github.com/timtebeek/launchedurlclassloader-illegalaccesserror
With packaged JarLauncher according to: https://spring.io/blog/2018/11/08/spring-boot-in-a-container
> We used a hard-coded main application class hello.Application. [...] You could also copy the Spring Boot fat JarLauncher into the image and use it to run the app - it would work and you wouldn’t need to specify the main class [...].

In my sample application `SomeRestController` contains a package private `SomeDataWrapperDTO` class, which can not be found in the _layered(!)_ Dockerized application. Without the JarLauncher, using an explicit main class, it works.

**Steps to reproduce**
1. `git clone git@github.com:timtebeek/launchedurlclassloader-illegalaccesserror.git`
2. `./mvnw verify && docker build -t lurlcl . && docker run -it --rm -p 8080:8080 lurlcl`
3. `http :8080/get`

**Expected**
HTTP 200 with `{"value": "some-value"}`

**Actual**
HTTP 500 with
```
{   "error": "Internal Server Error", 
    "message": "failed to access class com.github.timtebeek.SomeDataWrapperDTO from 
class com.github.timtebeek.SomeRestController$$EnhancerBySpringCGLIB$$a01253ed 
(com.github.timtebeek.SomeDataWrapperDTO is in unnamed module of loader 'app'; 
com.github.timtebeek.SomeRestController$$EnhancerBySpringCGLIB$$a01253ed is in unnamed 
module of loader org.springframework.boot.loader.LaunchedURLClassLoader @3cd1f1c8)", ... }
```

Note that the application works when the `@Validated` annotation is removed from the RestController; so it might be related to the cglib enhanced intermediate classes created.

**Comment by mbhave** [comment]
This appears to be because different classloaders being used to load the proxy for `SomeRestController` and `SomeDataWrapperDTO`. The `proxy` is loaded by the `LaunchedURLClassLoader` and `SomeDataWrapperDTO` is loaded by the app class loader. I think what @wilkinsona said in [this comment](https://github.com/spring-projects/spring-boot/issues/7267#issuecomment-310147172) is related although the actual issue is quite unrelated. 

> A package is scoped to the class loader that defined it so, in this case, we end up with two different `com.github.timtebeek` packages and a package-private type in one of those package is not visible to a type in the other of those packages.

Flagging for team attention to see what the rest of the team thinks.

**Comment by mbhave** [comment]
Also, the `javadoc` for the `JarLauncher` says

> This launcher assumes that dependency jars are included inside a /BOOT-INF/lib directory and that application classes are included inside a /BOOT-INF/classes directory. 

I wonder if it's a bit unusual to use the launcher once the libs have been moved out of `/BOOT-INF`.

**Comment by timtebeek** [comment]
Thanks for getting back to me on detail so quickly. It's near bedtime here, but I'll try moving the files around tomorrow morning and let you know how/if it helps. Since I'd also need to update the `-cp|` I'm not yet sure what the results will be.

**Comment by timtebeek** [comment]
> I wonder if it's a bit unusual to use the launcher once the libs have been moved out of /BOOT-INF.

You're absolutely right; After restoring the `BOOT-INF/` folder it works again: https://github.com/timtebeek/launchedurlclassloader-illegalaccesserror/commit/47532d243a21bad40bcf3bd35d6124f7575bcef4?diff=split

I'd built upon "A Better Dockerfile" in https://spring.io/blog/2018/11/08/spring-boot-in-a-container and merely added the `JarLauncher`; but as you've pointed out, when you use the `JarLauncher` one should also use `BOOT-INF/`. As a bonus I was able to drop the `-cp` argument.

Thanks a lot! Closing this issue. :)
